File: backend/app/schemas/analytics.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel

from app.schemas.validation import ValidationIssueCodeCount


class AnalyticsWarning(BaseModel):
    code: str
    message: str


class CalculationMethodRead(BaseModel):
    code: str
    name: str
    available: bool
    recommended: bool = False
    reason: str | None = None
    warning: str | None = None


class AnalyticsFieldsResponse(BaseModel):
    dimensions: list[str]
    metrics: list[str]
    aggregations: list[str]


class AnalyticsOverviewResponse(BaseModel):
    batch_id: int | None
    calculation_profile_id: int | None
    calculation_method: str | None = None
    calculation_method_name: str | None = None
    recommended_method: str | None = None
    statistics_algorithm: str | None = None
    statistics_label: str | None = None
    weight_source: str | None = None
    weight_count: int = 0
    weight_total: float | None = None
    is_weight_normalized: bool = False
    normalized_actual_progress: float | None = None
    normalized_planned_progress: float | None = None
    project_contribution_actual: float | None = None
    project_contribution_planned: float | None = None
    weight_warning: str | None = None
    recommendation_reason: str | None = None
    calculation_method_description: str | None = None
    available_methods: list[CalculationMethodRead] = []
    available_calculation_methods: list[CalculationMethodRead] = []
    mixed_units: bool = False
    unit_list: list[str] = []
    weight_sum: float | None = None
    included_batch_count: int = 0
    included_batches: list[ProjectOverviewBatch] = []
    baseline_plan_id: int | None = None
    baseline_plan_name: str | None = None
    batch_bound_baseline_plan_id: int | None = None
    batch_bound_baseline_plan_name: str | None = None
    current_view_baseline_plan_id: int | None = None
    current_view_baseline_plan_name: str | None = None
    baseline_consistent: bool = True
    baseline_notice: str | None = None
    item_count: int
    task_count: int
    actual_percent: float | None
    planned_percent: float | None
    actual_progress: float | None = None
    planned_progress: float | None = None
    time_planned_percent: float | None = None
    imported_planned_percent: float | None = None
    progress_deviation: float | None
    total_quantity: float | None
    actual_quantity: float | None
    remaining_quantity: float | None
    status_counts: dict[str, int]
    unit_mixed: bool = False
    warning: str | None = None
    warnings: list[AnalyticsWarning] = []
    batch_status_label: str | None = None
    batch_is_frozen: bool = False
    project_is_archived: bool = False


class ProjectOverviewBatch(BaseModel):
    batch_id: int
    sheet_name: str | None = None
    import_group_id: str | None = None
    import_group_name: str | None = None
    data_date: date | None = None
    actual_percent: float | None = None
    planned_percent: float | None = None
    item_count: int = 0


class ProjectOverviewResponse(BaseModel):
    project_actual_percent: float | None = None
    project_planned_percent: float | None = None
    project_deviation: float | None = None
    actual_progress: float | None = None
    planned_progress: float | None = None
    progress_deviation: float | None = None
    data_date: date | None = None
    included_batch_count: int = 0
    included_batches: list[ProjectOverviewBatch] = []
    calculation_method: str
    calculation_method_name: str | None = None
    recommended_method: str | None = None
    recommendation_reason: str | None = None
    available_methods: list[CalculationMethodRead] = []
    mixed_units: bool = False
    unit_list: list[str] = []
    statistics_label: str | None = None
    weight_sum: float | None = None
    item_count: int = 0
    task_count: int = 0
    is_project_aggregate: bool = True
    empty: bool = False
    message: str | None = None
    scope_label: str | None = None
    warning: str | None = None
    warnings: list[AnalyticsWarning] = []


class AnalyticsGroupRow(BaseModel):
    dimension_value: str | None
    value: float | int | None
    count: int
    units: list[str] = []
    unit_mixed: bool = False
    warning: str | None = None


class AnalyticsGroupByResponse(BaseModel):
    batch_id: int | None
    dimension: str
    metric: str
    aggregation: str
    rows: list[AnalyticsGroupRow]


class AnalyticsPlanVsActualRow(BaseModel):
    dimension_value: str | None
    actual_percent: float | None
    planned_percent: float | None
    time_planned_percent: float | None = None
    imported_planned_percent: float | None = None
    progress_deviation: float | None
    delayed_count: int = 0
    count: int
    units: list[str] = []
    unit_mixed: bool = False
    warning: str | None = None


class AnalyticsPlanVsActualResponse(BaseModel):
    batch_id: int | None
    dimension: str
    rows: list[AnalyticsPlanVsActualRow]


class DeviationAttributionRow(BaseModel):
    dimension_values: dict[str, str]
    actual_percent: float | None
    planned_percent: float | None
    progress_deviation: float | None
    abs_deviation: float | None
    delayed_count: int = 0
    count: int
    contribution: float | None = None
    warning: str | None = None


class DeviationAttributionResponse(BaseModel):
    batch_id: int | None
    dimensions: list[str]
    total_count: int
    overall_actual_percent: float | None = None
    overall_planned_percent: float | None = None
    overall_progress_deviation: float | None = None
    rows: list[DeviationAttributionRow]


class AnalyticsBuildingFloorItem(BaseModel):
    building: str
    floor: str
    task_count: int
    actual_percent: float | None
    planned_percent: float | None
    time_planned_percent: float | None = None
    imported_planned_percent: float | None = None
    progress_deviation: float | None
    delayed_count: int
    unit_mixed: bool = False
    units: list[str] = []
    warning: str | None = None


class AnalyticsBuildingFloorResponse(BaseModel):
    batch_id: int | None
    buildings: list[str]
    selected_building: str | None = None
    items: list[AnalyticsBuildingFloorItem]


class AnalyticsDelayedItem(BaseModel):
    id: int
    progress_item_id: int
    task_id: int | None
    task_name: str
    wbs_code: str | None
    task_code: str | None
    area: str | None
    construction_unit: str | None = None
    building: str
    floor: str
    discipline: str
    system_name: str
    unit: str | None
    actual_percent: float | None
    planned_percent: float | None
    time_planned_percent: float | None = None
    imported_planned_percent: float | None = None
    progress_deviation: float | None
    status: str | None
    delay_level: str
    delay_level_label: str
    delay_message: str
    rectification_item_id: int | None = None
    has_rectification: bool = False


class AnalyticsDelayedRankingResponse(BaseModel):
    batch_id: int | None
    rows: list[AnalyticsDelayedItem]


class AnalyticsTrendPoint(BaseModel):
    batch_id: int
    sheet_name: str | None = None
    status: str | None = None
    is_frozen: bool = False
    data_date: date | None
    published_at: datetime | None
    baseline_plan_id: int | None = None
    baseline_plan_name: str | None = None
    actual_percent: float | None
    planned_percent: float | None
    time_planned_percent: float | None = None
    imported_planned_percent: float | None = None
    progress_deviation: float | None
    schedule_phase: str | None = None
    item_count: int
    unit_mixed: bool = False
    warning: str | None = None


class AnalyticsTrendResponse(BaseModel):
    calculation_profile_id: int | None
    rows: list[AnalyticsTrendPoint]


class AnalyticsDataQualityResponse(BaseModel):
    batch_id: int | None
    data_quality_score: float | None
    field_completeness: float | None
    task_match_rate: float | None
    valid_row_rate: float | None
    plan_field_completeness: float | None
    unit_consistency: float | None
    warning_count: int
    error_count: int
    issue_code_counts: list[ValidationIssueCodeCount] = []
    status: str | None
    extra: dict[str, Any] = {}


class AnalyticsInsightResponse(BaseModel):
    overview_summary: str
    discipline_summary: str
    floor_summary: str
    building_floor_summary: str
    delay_summary: str
    quality_summary: str
    focus_points: list[str]
    recommended_actions: list[str]
    generated_at: str


class BaselineComparisonResponse(BaseModel):
    batch_id: int
    batch_bound_baseline_plan_id: int | None = None
    batch_bound_baseline_plan_name: str | None = None
    current_view_baseline_plan_id: int | None = None
    current_view_baseline_plan_name: str | None = None
    is_consistent: bool
    notice: str
    has_comparable_data: bool = True


class DashboardPlusDisciplineProgressRow(BaseModel):
    discipline: str
    task_count: int
    actual_percent: float | None = None
    planned_percent: float | None = None
    progress_deviation: float | None = None
    delayed_count: int = 0
    seriously_delayed_count: int = 0
    is_seriously_delayed: bool = False
    unit_mixed: bool = False
    units: list[str] = []
    warning: str | None = None


class DashboardPlusFloorDisciplineCell(BaseModel):
    floor: str
    discipline: str
    task_count: int
    actual_percent: float | None = None
    progress_deviation: float | None = None
    delayed_count: int = 0


class DashboardPlusBuildingDisciplineCell(BaseModel):
    building: str
    discipline: str
    task_count: int
    actual_percent: float | None = None
    delayed_count: int = 0


class DashboardPlusTaskDetail(BaseModel):
    id: int
    construction_unit: str | None = None
    building: str
    floor: str
    discipline: str
    task_name: str
    actual_percent: float | None = None
    planned_percent: float | None = None
    progress_deviation: float | None = None
    status: str | None = None
    delay_level: str
    delay_level_label: str


class DashboardPlusDelayStatusCount(BaseModel):
    status: str
    status_label: str
    count: int


class DashboardPlusDisciplineDelayCount(BaseModel):
    discipline: str
    seriously_delayed_count: int = 0
    delayed_count: int = 0
    slightly_delayed_count: int = 0
    normal_count: int = 0
    ahead_count: int = 0
    total_delayed_count: int = 0


class DashboardPlusDelayDistribution(BaseModel):
    status_counts: list[DashboardPlusDelayStatusCount] = []
    discipline_delay_counts: list[DashboardPlusDisciplineDelayCount] = []


class DashboardPlusResponse(BaseModel):
    batch_id: int | None = None
    filters: dict[str, str | None] = {}
    has_floor_data: bool = False
    has_building_data: bool = False
    discipline_progress: list[DashboardPlusDisciplineProgressRow] = []
    floor_discipline_matrix: list[DashboardPlusFloorDisciplineCell] = []
    building_discipline_matrix: list[DashboardPlusBuildingDisciplineCell] = []
    delay_distribution: DashboardPlusDelayDistribution = DashboardPlusDelayDistribution()
    task_details: list[DashboardPlusTaskDetail] = []


class DashboardUnifiedFilters(BaseModel):
    project_id: int
    data_date: date | None = None
    import_group_id: str | None = None
    batch_id: int | None = None
    sheet_name: str | None = None
    construction_unit: str | None = None
    building: str | None = None
    floor: str | None = None
    discipline: str | None = None
    system_name: str | None = None
    status: str | None = None
    calculation_method: str | None = None
    baseline_plan_id: int | None = None
    calculation_profile_id: int | None = None
    scope_label: str | None = None
    message: str | None = None


class DashboardUnifiedStatRow(BaseModel):
    name: str
    construction_unit: str | None = None
    building: str | None = None
    floor: str | None = None
    discipline: str | None = None
    system_name: str | None = None
    task_count: int = 0
    actual_percent: float | None = None
    planned_percent: float | None = None
    progress_deviation: float | None = None
    delayed_count: int = 0
    seriously_delayed_count: int = 0
    warning_count: int = 0
    rectification_count: int = 0
    calculation_method: str | None = None


class DashboardUnifiedMatrixRow(BaseModel):
    building: str | None = None
    floor: str | None = None
    discipline: str | None = None
    task_count: int = 0
    actual_percent: float | None = None
    planned_percent: float | None = None
    progress_deviation: float | None = None
    delayed_count: int = 0
    calculation_method: str | None = None


class DashboardBuildingElevationFloor(BaseModel):
    floor: str
    task_count: int = 0
    actual_percent: float | None = None
    planned_percent: float | None = None
    progress_deviation: float | None = None
    deviation: float | None = None
    delayed_count: int = 0
    serious_delayed_count: int = 0
    not_started_count: int = 0
    status: str = "no_data"
    status_label: str = "无数据"
    major_delayed_tasks: list[AnalyticsDelayedItem] = []
    calculation_method: str | None = None


class DashboardBuildingElevation(BaseModel):
    building: str
    floors: list[DashboardBuildingElevationFloor] = []
    task_count: int = 0
    actual_percent: float | None = None
    planned_percent: float | None = None
    progress_deviation: float | None = None
    status: str = "no_data"
    status_label: str = "无数据"
    message: str | None = None


class DashboardUnifiedDelayDistributionRow(BaseModel):
    status: str
    status_label: str
    count: int
    calculation_method: str | None = None


class DashboardUnifiedSummary(BaseModel):
    total: int = 0
    open: int = 0
    in_progress: int = 0
    completed: int = 0
    closed: int = 0
    overdue: int = 0
    unresolved: int = 0
    critical: int = 0
    warning: int = 0
    info: int = 0
    calculation_method: str | None = None


class DashboardUnifiedOptions(BaseModel):
    construction_units: list[str] = []
    buildings: list[str] = []
    floors: list[str] = []
    disciplines: list[str] = []
    systems: list[str] = []
    statuses: list[str] = []


class DashboardUnifiedResponse(BaseModel):
    filters: DashboardUnifiedFilters
    options: DashboardUnifiedOptions = DashboardUnifiedOptions()
    overview: AnalyticsOverviewResponse | None = None
    calculation_context: dict[str, Any] = {}
    by_construction_unit: list[DashboardUnifiedStatRow] = []
    by_building: list[DashboardUnifiedStatRow] = []
    by_floor: list[DashboardUnifiedStatRow] = []
    by_discipline: list[DashboardUnifiedStatRow] = []
    by_system: list[DashboardUnifiedStatRow] = []
    building_floor_matrix: list[DashboardUnifiedMatrixRow] = []
    building_elevation: list[DashboardBuildingElevation] = []
    discipline_floor_matrix: list[DashboardUnifiedMatrixRow] = []
    delay_distribution: list[DashboardUnifiedDelayDistributionRow] = []
    delayed_items: list[AnalyticsDelayedItem] = []
    warning_summary: DashboardUnifiedSummary = DashboardUnifiedSummary()
    rectification_summary: DashboardUnifiedSummary = DashboardUnifiedSummary()


class DashboardV2Scope(BaseModel):
    project_id: int
    view_mode: str = "overview"
    scope_label: str | None = None
    message: str | None = None
    filters: DashboardUnifiedFilters
    options: DashboardUnifiedOptions = DashboardUnifiedOptions()


class DashboardV2Response(BaseModel):
    scope: DashboardV2Scope
    overview: AnalyticsOverviewResponse | None = None
    discipline_cards: list[DashboardUnifiedStatRow] = []
    building_cards: list[DashboardUnifiedStatRow] = []
    floor_heatmap: list[DashboardUnifiedMatrixRow] = []
    building_elevation: list[DashboardBuildingElevation] = []
    delay_distribution: list[DashboardUnifiedDelayDistributionRow] = []
    delayed_items: list[AnalyticsDelayedItem] = []
    rectification_summary: DashboardUnifiedSummary = DashboardUnifiedSummary()
    calculation_context: dict[str, Any] = {}
    calculation_diagnostics: dict[str, Any] = {}
    dashboard_capabilities: dict[str, Any] = {}

File: frontend/src/types/analytics.ts
export type AnalyticsWarning = {
  code: string
  message: string
}

export type AnalyticsFieldsResponse = {
  dimensions: string[]
  metrics: string[]
  aggregations: string[]
}

export type CalculationMethodRead = {
  code: string
  name: string
  available: boolean
  recommended: boolean
  reason?: string | null
  warning?: string | null
}

export type AnalyticsOverviewResponse = {
  batch_id: number | null
  calculation_profile_id: number | null
  calculation_method?: string | null
  calculation_method_name?: string | null
  recommended_method?: string | null
  statistics_algorithm?: string | null
  statistics_label?: string | null
  weight_source?: string | null
  weight_count: number
  weight_total?: number | null
  is_weight_normalized: boolean
  normalized_actual_progress?: number | null
  normalized_planned_progress?: number | null
  project_contribution_actual?: number | null
  project_contribution_planned?: number | null
  weight_warning?: string | null
  recommendation_reason?: string | null
  calculation_method_description?: string | null
  available_methods?: CalculationMethodRead[]
  available_calculation_methods: CalculationMethodRead[]
  mixed_units?: boolean
  unit_list?: string[]
  weight_sum?: number | null
  included_batch_count?: number
  included_batches?: ProjectOverviewBatch[]
  baseline_plan_id?: number | null
  baseline_plan_name?: string | null
  batch_bound_baseline_plan_id?: number | null
  batch_bound_baseline_plan_name?: string | null
  current_view_baseline_plan_id?: number | null
  current_view_baseline_plan_name?: string | null
  baseline_consistent: boolean
  baseline_notice?: string | null
  item_count: number
  task_count: number
  actual_percent: number | null
  planned_percent: number | null
  actual_progress?: number | null
  planned_progress?: number | null
  time_planned_percent?: number | null
  imported_planned_percent?: number | null
  progress_deviation: number | null
  total_quantity: number | null
  actual_quantity: number | null
  remaining_quantity: number | null
  status_counts: Record<string, number>
  unit_mixed: boolean
  warning?: string | null
  warnings: AnalyticsWarning[]
  batch_status_label?: string | null
  batch_is_frozen: boolean
  project_is_archived: boolean
}

export type ProjectOverviewBatch = {
  batch_id: number
  sheet_name?: string | null
  import_group_id?: string | null
  import_group_name?: string | null
  data_date?: string | null
  actual_percent?: number | null
  planned_percent?: number | null
  item_count: number
}

export type ProjectOverviewResponse = {
  project_actual_percent?: number | null
  project_planned_percent?: number | null
  project_deviation?: number | null
  actual_progress?: number | null
  planned_progress?: number | null
  progress_deviation?: number | null
  data_date?: string | null
  included_batch_count: number
  included_batches: ProjectOverviewBatch[]
  calculation_method: string
  calculation_method_name?: string | null
  recommended_method?: string | null
  recommendation_reason?: string | null
  available_methods?: CalculationMethodRead[]
  mixed_units?: boolean
  unit_list?: string[]
  statistics_label?: string | null
  weight_sum?: number | null
  item_count: number
  task_count: number
  is_project_aggregate: boolean
  empty: boolean
  message?: string | null
  scope_label?: string | null
  warning?: string | null
  warnings: AnalyticsWarning[]
}

export type AnalyticsGroupRow = {
  dimension_value: string | null
  name?: string | null
  floor?: string | null
  value: number | null
  count: number
  task_count?: number
  actual_percent?: number | null
  planned_percent?: number | null
  progress_deviation?: number | null
  delayed_count?: number
  units: string[]
  unit_mixed: boolean
  warning?: string | null
}

export type AnalyticsGroupByResponse = {
  batch_id: number | null
  dimension: string
  metric: string
  aggregation: string
  rows: AnalyticsGroupRow[]
}

export type AnalyticsPlanVsActualRow = {
  dimension_value: string | null
  actual_percent: number | null
  planned_percent: number | null
  time_planned_percent?: number | null
  imported_planned_percent?: number | null
  progress_deviation: number | null
  delayed_count: number
  count: number
  units: string[]
  unit_mixed: boolean
  warning?: string | null
}

export type AnalyticsPlanVsActualResponse = {
  batch_id: number | null
  dimension: string
  rows: AnalyticsPlanVsActualRow[]
}

export type AnalyticsBuildingFloorItem = {
  building: string
  floor: string
  task_count: number
  actual_percent: number | null
  planned_percent: number | null
  time_planned_percent?: number | null
  imported_planned_percent?: number | null
  progress_deviation: number | null
  delayed_count: number
  unit_mixed: boolean
  units: string[]
  warning?: string | null
}

export type AnalyticsBuildingFloorResponse = {
  batch_id: number | null
  buildings: string[]
  selected_building?: string | null
  items: AnalyticsBuildingFloorItem[]
}

export type AnalyticsDelayedItem = {
  id: number
  progress_item_id: number
  task_id?: number | null
  task_name: string
  wbs_code?: string | null
  task_code?: string | null
  area?: string | null
  construction_unit?: string | null
  building: string
  floor: string
  discipline: string
  system_name: string
  unit?: string | null
  actual_percent?: number | null
  planned_percent?: number | null
  time_planned_percent?: number | null
  imported_planned_percent?: number | null
  progress_deviation?: number | null
  schedule_phase?: string | null
  status?: string | null
  delay_level: string
  delay_level_label: string
  delay_message: string
  rectification_item_id?: number | null
  has_rectification?: boolean
}

export type AnalyticsDelayedRankingResponse = {
  batch_id: number | null
  rows: AnalyticsDelayedItem[]
}

export type AnalyticsTrendPoint = {
  batch_id: number
  sheet_name?: string | null
  status?: string | null
  is_frozen: boolean
  data_date?: string | null
  published_at?: string | null
  baseline_plan_id?: number | null
  baseline_plan_name?: string | null
  actual_percent?: number | null
  planned_percent?: number | null
  progress_deviation?: number | null
  item_count: number
  unit_mixed: boolean
  warning?: string | null
}

export type AnalyticsTrendResponse = {
  calculation_profile_id: number | null
  rows: AnalyticsTrendPoint[]
}

export type AnalyticsDataQualityResponse = {
  batch_id: number | null
  data_quality_score?: number | null
  field_completeness?: number | null
  task_match_rate?: number | null
  valid_row_rate?: number | null
  plan_field_completeness?: number | null
  unit_consistency?: number | null
  warning_count: number
  error_count: number
  issue_code_counts: Array<{
    code: string
    level: string
    count: number
  }>
  status?: string | null
}

export type AnalyticsInsightResponse = {
  overview_summary: string
  discipline_summary: string
  floor_summary: string
  building_floor_summary: string
  delay_summary: string
  quality_summary: string
  focus_points: string[]
  recommended_actions: string[]
  generated_at: string
}

export type BaselineComparisonResponse = {
  batch_id: number
  batch_bound_baseline_plan_id?: number | null
  batch_bound_baseline_plan_name?: string | null
  current_view_baseline_plan_id?: number | null
  current_view_baseline_plan_name?: string | null
  is_consistent: boolean
  notice: string
  has_comparable_data: boolean
}

export type DashboardUnifiedFilters = {
  project_id: number
  data_date?: string | null
  import_group_id?: string | null
  batch_id?: number | null
  sheet_name?: string | null
  construction_unit?: string | null
  building?: string | null
  floor?: string | null
  discipline?: string | null
  system_name?: string | null
  status?: string | null
  calculation_method?: string | null
  baseline_plan_id?: number | null
  calculation_profile_id?: number | null
  scope_label?: string | null
  message?: string | null
}

export type DashboardUnifiedStatRow = {
  name: string
  construction_unit?: string | null
  building?: string | null
  floor?: string | null
  discipline?: string | null
  system_name?: string | null
  task_count: number
  actual_percent?: number | null
  planned_percent?: number | null
  progress_deviation?: number | null
  delayed_count: number
  seriously_delayed_count: number
  warning_count: number
  rectification_count: number
  calculation_method?: string | null
}

export type DashboardUnifiedMatrixRow = {
  building?: string | null
  floor?: string | null
  discipline?: string | null
  task_count: number
  actual_percent?: number | null
  planned_percent?: number | null
  progress_deviation?: number | null
  delayed_count: number
  calculation_method?: string | null
}

export type DashboardBuildingElevationFloor = {
  floor: string
  task_count: number
  actual_percent?: number | null
  planned_percent?: number | null
  progress_deviation?: number | null
  deviation?: number | null
  delayed_count: number
  serious_delayed_count: number
  not_started_count: number
  status: string
  status_label: string
  major_delayed_tasks: AnalyticsDelayedItem[]
  calculation_method?: string | null
}

export type DashboardBuildingElevation = {
  building: string
  floors: DashboardBuildingElevationFloor[]
  task_count: number
  actual_percent?: number | null
  planned_percent?: number | null
  progress_deviation?: number | null
  status: string
  status_label: string
  message?: string | null
}

export type DashboardUnifiedSummary = {
  total: number
  open: number
  in_progress: number
  completed: number
  closed: number
  overdue: number
  unresolved: number
  critical: number
  warning: number
  info: number
  calculation_method?: string | null
}

export type DashboardUnifiedResponse = {
  filters: DashboardUnifiedFilters
  options: {
    construction_units: string[]
    buildings: string[]
    floors: string[]
    disciplines: string[]
    systems: string[]
    statuses: string[]
  }
  overview?: AnalyticsOverviewResponse | null
  calculation_context: Record<string, unknown>
  by_construction_unit: DashboardUnifiedStatRow[]
  by_building: DashboardUnifiedStatRow[]
  by_floor: DashboardUnifiedStatRow[]
  by_discipline: DashboardUnifiedStatRow[]
  by_system: DashboardUnifiedStatRow[]
  building_floor_matrix: DashboardUnifiedMatrixRow[]
  building_elevation: DashboardBuildingElevation[]
  discipline_floor_matrix: DashboardUnifiedMatrixRow[]
  delay_distribution: Array<{ status: string; status_label: string; count: number; calculation_method?: string | null }>
  delayed_items: AnalyticsDelayedItem[]
  warning_summary: DashboardUnifiedSummary
  rectification_summary: DashboardUnifiedSummary
}

export type DashboardV2Scope = {
  project_id: number
  view_mode: 'overview' | 'discipline' | 'building'
  scope_label?: string | null
  message?: string | null
  filters: DashboardUnifiedFilters
  options: DashboardUnifiedResponse['options']
}

export type DashboardV2Response = {
  scope: DashboardV2Scope
  overview?: AnalyticsOverviewResponse | null
  discipline_cards: DashboardUnifiedStatRow[]
  building_cards: DashboardUnifiedStatRow[]
  floor_heatmap: DashboardUnifiedMatrixRow[]
  building_elevation: DashboardBuildingElevation[]
  delay_distribution: Array<{ status: string; status_label: string; count: number; calculation_method?: string | null }>
  delayed_items: AnalyticsDelayedItem[]
  rectification_summary: DashboardUnifiedSummary
  calculation_context: Record<string, unknown>
  calculation_diagnostics: Record<string, unknown>
  dashboard_capabilities: Record<string, { available: boolean; reason: string }>
}
